### ISSUE TITLE
datapath: Get rid of DIRECT_ROUTING_DEV_IFINDEX, IPV{4,6}_DIRECT_ROUTING

### ISSUE THREAD

**Issue by brb** [cilium/cilium#17441]
I'm currently working on relaxing the `bpf_{xdp,skb}_fib_lookup()`, so that it can be used to determine ~`ifindex` of outgoing device and~ the src IP addr of it (will add the link to LORE once I've upstreamed it). This will allow us to get rid of the macros in the datapath, and will make multi-homed k8s possible.

**Comment by brb** [comment]
Crude PR for NodePort in the SNAT BPF mode to test the kernel changes - https://github.com/cilium/cilium/pull/28283

**Comment by brb** [comment]
The kernel patch was merged - https://lore.kernel.org/bpf/169689542525.10639.4943696470893944027.git-patchwork-notify@kernel.org/

**Comment by brb** [comment]
The kernel patch is currently being backported to 6.6, 6.1 and 5.15 stable. We should keep the fallback for 5.10 and 5.4 though.

**Comment by shouri007** [comment]
Hi @brb , do you have any update on the kernel patch backport?

**Comment by brb** [comment]
Hi @shouri007 , please see https://github.com/cilium/cilium/issues/17441#issuecomment-1972946953.

**Comment by aditighag** [comment]
Here is an overview of proposed changes in the datapath for dynamic source IP selection, moving away from hard-coded direct routing configurations:

1. Source IP Retrieval
Use `bpf_fib_lookup` with `BPF_FIB_LOOKUP_SRC` flag to dynamically retrieve a source IP address. 

2.  Removal of `ENABLE_SKIP_FIB`
The datapath currently [skips](https://github.com/cilium/cilium/blob/main/bpf/lib/fib.h#L168-L168) FIB lookup when there is only a single device on a host as an optimization. Get rid of the macro as the agent automatically attaches BPF programs to all detected devices on a host with the auto-detect feature. 

3. Tracking Conflicting Host Connections
`nodeport_has_nat_conflict_ipv{4,6}` [tracks](https://github.com/cilium/cilium/blob/main/bpf/lib/nodeport_egress.h#L287-L287) host originated connections with source IP set to `IPV{4,6}_DIRECT_ROUTING` that conflict with nodeport LB traffic DNAT and SNAT'd routed to remote backends. With the dynamic selection of source IP, such conflicting connections are tricky to track. While SNAT of such connections isn't necessary with the dynamic selection, @julianwiedmann suggested tracking all host originated traffic within the nodeport port range as a first step.

4. XDP Tunnel Headers
The agent currently [requires](https://github.com/cilium/cilium/blob/7ecc741633179b29a181ab3fcc60c5d6220e77bc/pkg/option/config.go#L2203-L2203) direct routing devices for manually building the tunnel headers for the XDP nodeport case -
.`IPV4_DIRECT_ROUTING` is used to set the source IP address in the outer tunnel header in `ctx_set_encap_info4`. This can be preceded with `bpf_fib_lookup` with `BPF_FIB_LOOKUP_SRC` flag in order to dynamically retrieve a source IP address for the outer tunnel header. 
@julianwiedmann pointed out that there is currently no support for manually building IPv6 tunnel header for the XDP case.

**Comment by julianwiedmann** [comment]
> 2\. Removal of `ENABLE_SKIP_FIB`
>         The datapath currently [skips](https://github.com/cilium/cilium/blob/main/bpf/lib/fib.h#L168-L168) FIB lookup when there is only a single device on a host as an optimization. Get rid of the macro as the agent automatically attaches BPF programs to all detected devices on a host with the auto-detect feature.

fwiw I'd suggest to already queue this change up for `v1.19`, and see if there's any fallout. It's essentially obsoleted by the device auto-detection, and it's hard to imagine in what scenario users would see a behavioral difference when it's removed.

**Comment by julianwiedmann** [comment]
I think we'll want a separate issue to also track the switch from a single hard-coded per-interface MASQ IP (`nat_ipv4_masquerade`), to instead using the sourceIP from the FIB lookup.

**Comment by julianwiedmann** [comment]
To me it looks like we've explored the N/S LB forwarding path in depth, and what changes are needed there :+1:.

I think it would be good to have the same picture for
1. pod-egressing traffic (with BPF Host-Routing enabled) - where/how does it now pick its egress interface and source IP?
2. N/S reply traffic handled by the LB node (ie `nodeport_rev_dnat_ipv4()`) - what impact do we expect here, if any?

**Comment by aditighag** [comment]
Preparatory PR to remove ENABLE_SKIP_FIB - [PR #43277](https://github.com/cilium/cilium/pull/43277/commits)

> I think we'll want a separate issue to also track the switch from a single hard-coded per-interface MASQ IP (`nat_ipv4_masquerade`), to instead using the sourceIP from the FIB lookup.

That's right, I think we already have one - https://github.com/cilium/cilium/issues/17158.

>pod-egressing traffic (with BPF Host-Routing enabled) - where/how does it now pick its egress interface and source IP?

Based on my reviewing of the code, I hadn't noticed any particular paths to tackle in the lxc datapath in the context of this issue involving direct routing outside of the proposed changes above. 

Let's double check:
- pod to local pod, host: FIB lookups are straightforward with the source ip derived from the packet ctx. Packets are redirected based on the FIB results.
- pod to remote node: Traffic can be masqueraded in some cases, ref: 17158. There is a comment about traffic being tunneled in case of host fw, which isn't relevant to this discussion. 
- pod to nodeport: See below. 

*We would have to closely examine the endpoint routes code for any dependencies. I have also left out encryption code paths.

Do you have any specific network paths in mind? 

>N/S reply traffic handled by the LB node (ie nodeport_rev_dnat_ipv4()) - what impact do we expect here, if any?

This is covered by the proposed changes (1), essentially `nodeport_egress.h`. Feel free to call out if you see anything else.

**Comment by brb** [comment]
>  Use bpf_fib_lookup with BPF_FIB_LOOKUP_SRC flag to dynamically retrieve a source IP address.

Just keep in mind that the flag is available only since the stable 5.15 (unfortunately, not in the currently supported 5.10).

**Comment by julianwiedmann** [comment]
> > pod-egressing traffic (with BPF Host-Routing enabled) - where/how does it now pick its egress interface and source IP?
> 
> Based on my reviewing of the code, I hadn't noticed any particular paths to tackle in the lxc datapath in the context of this issue involving direct routing outside of the proposed changes above.
> 

Yep I think the big change is the removal of `ENABLE_SKIP_FIB`, so that pod-egressing traffic [here](https://github.com/cilium/cilium/blob/c43e8e9f66d4f97a1029b34c6117acf11ce8bd97/bpf/bpf_lxc.c#L1459) is exposed to the full spectrum of configured routing, and egress interfaces on the node. This gets interesting once we tackle the source IP selection for Masquerading ... will we have *two* FIB lookups (one in `bpf_lxc` for the egress interface, one in `to-netdev` for the source IP) 
> 
> > N/S reply traffic handled by the LB node (ie nodeport_rev_dnat_ipv4()) - what impact do we expect here, if any?
> 
> This is covered by the proposed changes (1), essentially `nodeport_egress.h`. Feel free to call out if you see anything else.

I was thinking of [this](https://github.com/cilium/cilium/blob/c43e8e9f66d4f97a1029b34c6117acf11ce8bd97/bpf/lib/nodeport.h#L2404) FIB-lookup based redirect, when we push a reply back to the cluster-external client. That will also be impacted by the `ENABLE_SKIP_FIB` removal, but there's no need to select any sourceIP (since the RevDNAT mechanism gives us the desired sourceIP).

**Comment by borkmann** [comment]
(related https://github.com/cilium/cilium/issues/39121 )

**Comment by ldelossa** [comment]
@julianwiedmann I am going to pick up on 

> Tracking Conflicting Host Connections
nodeport_has_nat_conflict_ipv{4,6} [tracks](https://github.com/cilium/cilium/blob/main/bpf/lib/nodeport_egress.h#L287-L287) host originated connections with source IP set to IPV{4,6}_DIRECT_ROUTING that conflict with nodeport LB traffic DNAT and SNAT'd routed to remote backends. With the dynamic selection of source IP, such conflicting connections are tricky to track. While SNAT of such connections isn't necessary with the dynamic selection, @julianwiedmann suggested tracking all host originated traffic within the nodeport port range as a first step.

If we are moving to multi-homed Cilium nodes, we can no longer depend on the host-networking check done in `nodeport_has_nat_conflict_ipv` to avoid sNAT conflicts, as this assumes:
1. host generated traffic always egresses via the direct routing interface.
2. host generated traffic always binds to the ip address of the direct routing interface.

Moving forward the nodeport code will need to identify host networking traffic from multiple IPs on multiple egress interfaces. 
I believe we can now check `ctx->ingress_interface` in the egress nodeport/sNAT path to determine if the skb is locally generated. I looked at the TCP, UDP, and SCTP paths in the kernel and AFAICT the local out paths never set `sk_buff->skb_iif`. 

This should allow the nodeport code to reserve a NAT port or translate host traffic to avoid port conflicts, regardless of the interface the host network traffic egresses.

**Comment by julianwiedmann** [comment]
> I believe we can now check `ctx->ingress_interface` in the egress nodeport/sNAT path to determine if the skb is locally generated. I looked at the TCP, UDP, and SCTP paths in the kernel and AFAICT the local out paths never set `sk_buff->skb_iif`.

How would that look like for pod-originating traffic though? There we don't have to worry about source-IP clashes with Nodeport-forwarded traffic. And definitely don't want to apply NAT-tracking for all pod connections.

**Comment by ldelossa** [comment]
@julianwiedmann that should not occur.

In the veth's xmit path we wind up on the given CPU's backlog where the ordinary [`netif_recieve_skb` path](https://elixir.bootlin.com/linux/latest/source/drivers/net/veth.c#L321) is invoked.

In the dequeue path [we always set skb_iif prior to running TC/TCX programs](https://elixir.bootlin.com/linux/latest/source/net/core/dev.c#L5881), meaning even if we redirect before seeing the linux net stack, skb_iif will be set as the input interface for pod-originating traffic. Likewise, I don't see anything in the forwarding routing path which scrubs or sets the `skb_iif` field, so it should remain in legacy routing mode as well. 

There may of course be some edge cases here that I'm not aware of.

**Comment by borkmann** [comment]
> [@julianwiedmann](https://github.com/julianwiedmann) that should not occur.
> 
> In the veth's xmit path we wind up on the given CPU's backlog where the ordinary [`netif_recieve_skb` path](https://elixir.bootlin.com/linux/latest/source/drivers/net/veth.c#L321) is invoked.
> 
> In the dequeue path [we always set skb_iif prior to running TC/TCX programs](https://elixir.bootlin.com/linux/latest/source/net/core/dev.c#L5881), meaning even if we redirect before seeing the linux net stack, skb_iif will be set as the input interface for pod-originating traffic. Likewise, I don't see anything in the forwarding routing path which scrubs or sets the `skb_iif` field, so it should remain in legacy routing mode as well.
> 
> There may of course be some edge cases here that I'm not aware of.

For the case of bpf_lxc, the skb_iif field will be populated in case of regular veth/tcx when packets going given the hook is at tcx ingress for egress traffic. In case of netkit, the skb_iif field will /not/ be populated since the bpf_lxc is executed inside the netkit peer driver (somewhat equivalent to tcx egress at veth "inside" the Pod). But this is mainly for traffic intending to be pushed out of the node. If the Pod with netkit targets traffic to go up the host stack, then the skb_iif field will be populated.

**Comment by ldelossa** [comment]
Hmm @borkmann 

This means however if the following occurs:

`pod` --> `netkit_peer` --bpf_redirect--> `eth0`  (which could be the case with eBPF host routing right?) 

Then we'd hit nodeport/sNAT code with an unset `skb_iif`? 

Also 
>  If the Pod with netkit targets traffic to go up the host stack, then the skb_iif field

Can you elaborate on this? Do you mean here, just that `bpf_lxc` is dropping to stack?

**Comment by borkmann** [comment]
> Hmm [@borkmann](https://github.com/borkmann)
> 
> This means however if the following occurs:
> 
> `pod` --> `netkit_peer` --bpf_redirect--> `eth0` (which could be the case with eBPF host routing right?)

Yes.

> Then we'd hit nodeport/sNAT code with an unset `skb_iif`?

Yes, skb_iif = 0.

> > If the Pod with netkit targets traffic to go up the host stack, then the skb_iif field
> 
> Can you elaborate on this? Do you mean here, just that `bpf_lxc` is dropping to stack?

If the Pod want to talk to sth in the hostns, then netkit will act similarly as veth by looping the packet onto the netkit primary via ingress path. Then skb_iif should point to netkit primary device.

**Comment by julianwiedmann** [comment]
fwiw, I'm a bit surprised that we need such detailed knowledge about kernel internals. And would prefer not to debug it in the future ;).

Are we sure that this aspect can't be solved with Cilium-internal pieces - say with an endpoint lookup, and/or `MARK_MAGIC_HOST`?

**Comment by borkmann** [comment]
> Are we sure that this aspect can't be solved with Cilium-internal pieces - say with an endpoint lookup, and/or `MARK_MAGIC_HOST`?

Yeah if we can encode sth into the mark (or somewhere else), that would be good. We do know that a packet must have passed through the bpf_lxc program when coming from a Pod, no matter which device flavor.

**Comment by ldelossa** [comment]
It would have been nice to simply look at ingress interface, and say "okay there is no ingress interface, this is local traffic", less about thinking about kernel internals. But yes, with netkit this won't work. 

> Are we sure that this aspect can't be solved with Cilium-internal pieces - say with an endpoint lookup, and/or MARK_MAGIC_HOST?

Yeah I think so.
In the common case we mark locally generated traffic with `MagicMarkHost` via IPTables. 
We can detect this via `HOST_ID` source identity and perform the SNAT reservation when we see this. 

The important new distinction we need to make is forwarded traffic. Unless we keep a list of host source IPs, special attention must be taken to distinguish forwarded traffic vs local output traffic. We don't want to create SNAT reservations for forwarded traffic by mistake. 

That IPTable rule is obviously ignored for l7 egress proxy traffic, which can be transparent (sourced as pod) or not (sourced as host ip). 
We'll then deduce if the source is a host IP most likely from an endpoint map lookup. 
If it uses a host IP, we'll do the SNAT reservation. 

There seems to be a flag which disables IPTable rules in Cilium, which I need to explore to determine if we need another case, where nothing is marking local traffic at iptables.

**Comment by julianwiedmann** [comment]
> The important new distinction we need to make is forwarded traffic. Unless we keep a list of host source IPs, special attention must be taken to distinguish forwarded traffic vs local output traffic. We don't want to create SNAT reservations for forwarded traffic by mistake.

This being N/S traffic that's forwarded by the LB to a remote backend? I remember us setting the `SNAT_DONE` mark for such traffic, but not sure whether that's still well maintained. Either way, if it's forwarded then it definitely doesn't have the `MARK_MAGIC_HOST`.
 
> There seems to be a flag which disables IPTable rules in Cilium, which I need to explore to determine if we need another case, where nothing is marking local traffic at iptables.

I'd rather we just fatal if that flag is set 😁 .

**Comment by ldelossa** [comment]
@julianwiedmann okay great!

If we are okay with forcing IPTable rules when nodeport code is present, that simplifies things quite a bit. 
This should mean we only need to check for the `MARK_MAGIC_HOST` and the egress proxy marks. 

Like mentioned above, we'll just need some special handling to deduce if the egress proxy traffic is transparent or not, where in the former case we'll need to do a SNAT reservation. 

> This being N/S traffic that's forwarded by the LB to a remote backend? 

I was thinking more generically, since the host could see non-Cilium related traffic being forwarded between interfaces and we wouldn't want to create SNAT reservations for this. However, if we don't have a case where IPtables is not marking locally created traffic anymore, this becomes a moot point. 

Here is PR for this concept: https://github.com/cilium/cilium/pull/43955

**Comment by farodin91** [comment]
@ldelossa Out of curiosity, will this solve such an issue: https://github.com/cilium/cilium/discussions/40693 

Current host setup:
https://github.com/cilium/cilium/pull/40705#issuecomment-3121581225

**Comment by ldelossa** [comment]
https://github.com/cilium/cilium/pull/43955 has merged 🚀 

Source IP Retrieval
Use bpf_fib_lookup with BPF_FIB_LOOKUP_SRC flag to dynamically retrieve a source IP address.

~~Removal of ENABLE_SKIP_FIB
The datapath currently [skips](https://github.com/cilium/cilium/blob/main/bpf/lib/fib.h#L168-L168) FIB lookup when there is only a single device on a host as an optimization. Get rid of the macro as the agent automatically attaches BPF programs to all detected devices on a host with the auto-detect feature.~~

~~Tracking Conflicting Host Connections
nodeport_has_nat_conflict_ipv{4,6} [tracks](https://github.com/cilium/cilium/blob/main/bpf/lib/nodeport_egress.h#L287-L287) host originated connections with source IP set to IPV{4,6}_DIRECT_ROUTING that conflict with nodeport LB traffic DNAT and SNAT'd routed to remote backends. With the dynamic selection of source IP, such conflicting connections are tricky to track. While SNAT of such connections isn't necessary with the dynamic selection, @julianwiedmann suggested tracking all host originated traffic within the nodeport port range as a first step.~~

XDP Tunnel Headers
The agent currently [requires](https://github.com/cilium/cilium/blob/7ecc741633179b29a181ab3fcc60c5d6220e77bc/pkg/option/config.go#L2203-L2203) direct routing devices for manually building the tunnel headers for the XDP nodeport case -
.IPV4_DIRECT_ROUTING is used to set the source IP address in the outer tunnel header in ctx_set_encap_info4. This can be preceded with bpf_fib_lookup with BPF_FIB_LOOKUP_SRC flag in order to dynamically retrieve a source IP address for the outer tunnel header.
@julianwiedmann pointed out that there is currently no support for manually building IPv6 tunnel header for the XDP case.

I will begin looking at `XDP Tunnel Headers` next.

**Comment by ldelossa** [comment]
https://github.com/cilium/cilium/pull/44625 opened to handle the final items above.

**Comment by msune** [comment]
> https://github.com/cilium/cilium/pull/44625 opened to handle the final items above.

So I guess this is complete now, right?

**Comment by ldelossa** [comment]
@msune I don't think so, there are still other usages of the direct routing interfaces which need to be handled. RevDNAT is currently on my list. Its taking a bit of a back seat since the work done here so far, has unblocked us on other tasks of higher priority, but I will be returning to this ticket to outline the RevDNAT changes next.
